Add Q&A: securing data at rest, and FBE vs FDE

Peter-Easton opened the pull request:
Short little part about the encryption. This part is a little lengthy so I completely understand if it's not desired or it might do better in a separate page.

Peter-Easton committed:
Add Q&A: securing data at rest, and FBE vs FDE

Update taking into account behaviour observed when trying pincodes

Expand Q&A on advantages of FBE and clarify advantages of FBE

Clarify user profiles for Question about FBE vs FDE

Clarify Q&A about securing data at rest with numbers of how long it
would take to brute force the pincodes with the proper timeout
thresholds of the Titan M obtained by process of experimentation.

inthewaves commented on static/faq.html:
```diff
@@ -291,6 +294,55 @@ <h2 id="default-connections">
+                <a href="#securing-data-at-rest">How does GrapheneOS secure data at rest?</a>
+            </h2>
+            <p>GrapheneOS makes full use of the hardware-bound encryption in modern Android devices.
+            The Titan M, a specially designed hardware security module on Pixel 3 handsets handles
```
"HSM" is used below without any introduction, so we should introduce it here. Also, the Titan M is present on the 3a
```suggestion
            The Titan M, a specially designed hardware security module (HSM) on Pixel 3 and 3a handsets, handles
```

inthewaves commented on static/faq.html:
```diff
@@ -291,6 +294,55 @@ <h2 id="default-connections">
+            progressively lengthening delays between guesses at the password. On enough failed
+            guesses, the Titan M can increase the time between guesses to allow only one guess
+            per day. At this rate, to search through all the combinations of a four-digit pin code
+            from 0000 to 9950 would take over 27 years, or for a relatively weak two-word passphrase
```
Why 9950 and not 9999?

Peter-Easton replied:
Good thing you brought this up, since after you mentioned it I decided I'd give it a test using my own phone to see if it actually took more than 50 to make it from 30 seconds to a minute and up and up. Originally I was under the impression that it would behave by doubling the time interval, but I decided to experiment a little bit and test with a password, pincode, and pattern, and my results actually differed a bit from what I expected:

After the first five attempts the phone will display the number of incorrect password attempts in a dialogue box, disable the fingerprint sensor and reject all input for 30 seconds. After the next five attempts, the phone will display the number of incorrect password attempts in a dialogue box and reject all input for another 30 seconds. After each unsuccessful attempt at that point, it'll do the same.

I'll actually have to revisit this section, whoops.

thestinger replied:
The Titan M is only involved for the initial unlock after rebooting. Make sure you're properly testing this.

Peter-Easton replied:
I followed the instructions and tested it to 61 results this time. The time delay starts to increment from 30 seconds to 60 seconds at 40 tries, after which the time lockout doubles to two minutes at 50 tries, then four minutes at 60 tries. I think that provided that it continues to work this way, then it would take 100 tries to reach one hour, and then from there after 150 tries it would reach one day between attempts. 

While I was at it, I also plugged a keyboard into my phone and attempted to enter the password, and it looks like GrapheneOS won't allow this which would make automated password spoofing even more difficult. 

Looks like it is working exactly as it is intended to, I just was mistaken and thought that it would be incrementing faster. I'll update the draft with this new information.

thestinger replied:
It allows attaching a keyboard but only detects it at boot or when unlocked, by default.

inthewaves commented on static/faq.html:
```diff
@@ -291,6 +294,55 @@ <h2 id="default-connections">
+            timer which cannot be accessed from the host operating system, and can enforce
+            progressively lengthening delays between guesses at the password. On enough failed
+            guesses, the Titan M can increase the time between guesses to allow only one guess
+            per day. At this rate, to search through all the combinations of a four-digit pin code
```
pin code seems redundant
```suggestion
            per day. At this rate, to search through all the combinations of a four-digit PIN
```

inthewaves commented on static/faq.html:
```suggestion
            <p>Virtually all modern phones use file-based encryption instead of full-disk
```

inthewaves commented on static/faq.html:
```suggestion
            encryption. Android 10 requires file-based encryption. Utilizing file-based encryption
```

inthewaves commented on static/faq.html:
```diff
@@ -291,6 +294,55 @@ <h2 id="default-connections">
+            about the size of a grain of rice and is difficult to attack. Each file is encrypted
+            with AES-256-XTS, and the filesystem metadata is encrypted with AES-256-HEH.</p>
+
+            <p>File Based Encryption on GrapheneOS is always on, enforced, and cannot be switched
```
```suggestion
            <p>File-based encryption on GrapheneOS is always on, enforced, and cannot be switched
```

inthewaves commented on static/faq.html:
```suggestion
                    <li>File-based encryption allows some features like direct boot, which many
```

inthewaves commented on static/faq.html:
I think this can be made more general as it applies also for user profiles, not just work profiles

Peter-Easton replied:
Sure. Let me go and fix that...

Peter-Easton replied:
Okay! I dropped the word "work" profiles, how does it sound?

inthewaves replied:
I think it should be "user profiles", since the OS uses "profile" in the strings (and also https://source.android.com/devices/tech/admin/multi-user). "Profiles" alone may be ambiguous

I think it would also be good to mention per-profile encryption keys as well, since it's not explicitly mentioned here. 

So, something like, "It allows user profiles to be made more secure, since each profile has their own encryption keys that get wiped from memory when logged out. Data in an inactive profile becomes at rest (i.e., the data becomes inaccessible, because the encryption key has to be reobtained from user authentication)." This would also clarify what we mean by "at rest."

There's more detail given recently on this topic: https://reddit.com/r/GrapheneOS/comments/f1b2l7/wipe_free_space/fh8lvk9?context=3

Peter-Easton replied:
Sorry about that. I had some problems and this issue fell through the cracks. I'll fix that next push.

Peter-Easton replied:
Should be fixed now.

inthewaves commented on static/faq.html:
```diff
@@ -291,6 +294,55 @@ <h2 id="default-connections">
+            </h2>
+            <p>GrapheneOS makes full use of the hardware-bound encryption in modern Android devices.
+            The Titan M, a specially designed hardware security module on Pixel 3 handsets handles
+            validation of the user's pincode or password. Upon getting the correct password, the HSM
```
```suggestion
            validation of the user's PIN or password. Upon getting the correct password, the HSM
```

inthewaves commented on static/faq.html:
```diff
@@ -291,6 +294,55 @@ <h2 id="default-connections">
+            extract the SSD and connect it to a much more powerful system to run guessing attacks on
+            it offline, and instead must go through the specific HSM on the phone. The HSM is
+            about the size of a grain of rice and is difficult to attack. Each file is encrypted
+            with AES-256-XTS, and the filesystem metadata is encrypted with AES-256-HEH.</p>
```
The vocab is a bit inconsistent; it jumps between "HSM" in this paragraph to "Titan M" in the next paragraph

Peter-Easton replied:
Sure. I'll go and get that...

Peter-Easton replied:
Should be fixed now!

Peter-Easton committed:
Add time estimate to brute-force pattern unlock to FAQ

Revert "Add time estimate to brute-force pattern unlock to FAQ"

This reverts commit accfee283c4bd8149683d4fc777b98cd43699d92.

Add clarification about android lock pattern and caveat that human beings don't use it well.
